File: programmers/42586.cpp
/*
- 날짜 count
- 매 날짜마다 해당 speeds들 추가
- 100이상인 것들 pop(0) <- 이거 너무 오버헤드가 크다. 
    파이썬에선 디큐로 만들면 된다지만 c++에선 무엇을 활용해야할지 ?!
- 이때 한 번에 팝되는 얘들의 개수를 넣으면 된다.
    
[ 내 생각 ]
- 완료된 일을 모아놓는 벡터를 만들고 그 합이 기존 길이랑 같으면 리턴

-> 마지막 테케에서 걸린다. 100 이상인데, 100 초과로 설정해놔서 그렇다.
*/
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
        
    int count = 0;
    int start = 0;

    while(count != progresses.size()){
        
        int end = 0;
        for(int i = start ; i < progresses.size() ; i++){
            progresses[i] += speeds[i];
            
            // 작업 후 하루의 끝 ( 100이 넘었는가? 판단 )
            if( i == start && progresses[i] >= 100){
                end++;
                start++;
                count++;
            }
        }
        if(end !=0){
            answer.push_back(end);
        }
    }
    return answer;
}

File: programmers/42586.py
"""
- 해당 기능이 배포되는데 걸리는 날짜를 계산하는 방식 활용해보기 (큐 이용)
- 남은 날짜를 계산하는 과정에서 틀렸었다.
"""
from collections import deque
import math
def solution(progresses, speeds):
    answer = []
    
    # 기능 개발하는데 필요한 일 수 <- 여기서 틀렸었다.
    q = deque()
    for i in range(len(progresses)):
        day = (100 - progresses[i]) / speeds[i]
        if day % 1 != 0:
            day = math.floor(day) + 1
        q.append(int(day))
    
    day = q.popleft()
    count = 1
    while(q):
        temp = q.popleft()
        if(day < temp):
            answer.append(count)
            day = temp
            count = 1
        else:
            count += 1
    answer.append(count)
    return answer